### PR TITLE
Exthost - DocumentContentProvider, Decorations, Diagnostics

### PR DESCRIPTION
**Pull request by bryphe**
Implement several other API notifications for the main thread:

- DocumentContentProvider
- Decorations
- Diagnostics

And add a test case for diagnostics.

### CODE REVIEW

File: src/Core/Utility/ListEx.re [hunk 1]
```diff
@@ -11,6 +11,23 @@ let rec last =
...
+             | Ok(v) => Ok([v, ...list])
+             | Error(_) as err => err
+             }
+           })
+      },
+      Ok([]),
+      items,
+    );
+  };
+
```
Comment (glennsl) [inline src/Core/Utility/ListEx.re:30]: This is a pretty bad name, as it doesn't actually do any mapping from he user POV. It also belongs in `ResultEx`, not `ListEx`.  It's a tricky one to name though. In Haskell and Scalaz it's called `sequence`, in Elm it's called `combine`, in ocaml-containers it's called `flatten` and in Base it's called `all`.

That last one should serve as a hint that you don't even have to implement and name it yourself, but can just use `Base.Result.all`.

File: src/Exthost/Diagnostic.re [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      );
+
+    Ok({
+      range,
+      message: ret.message,
+      severity: ret.severity,
+      // TODO:
+      //source: ret.source,
+      //code: ret.code,
+    });
```
Comment (glennsl) [inline src/Exthost/Diagnostic.re:44]: Why not use `ocaml-decoders` and avoid all this unnecessary conversion and duplicate types? IMO `deriving yojson` should be outright banned.

File: src/Core/Utility/ListEx.re (hunk 1 above)
Comment (bryphe) [reply src/Core/Utility/ListEx.re:30]: Ah, cool - I figured it must exist somewhere.  I keep forgetting that we have `Base` around... would be nice if there was something like `hoogle` for all our dependencies, to search by function signature.

File: src/Exthost/Diagnostic.re (hunk 2 above)
Comment (bryphe) [reply src/Exthost/Diagnostic.re]: Ah, I ported this over from existing code... switched to use the decoder strategy now, though.

Comment (glennsl) [reply src/Exthost/Diagnostic.re]: Ah, sorry, from the description I thought it was new. But good to get it converted anyway!

**Pull request merged**